src/integration-test/groovy/org/pih/warehouse/component/auth/AuthApiSpec.groovy

```diff
@@ -0,0 +1,60 @@
+package org.pih.warehouse.component.auth
+
+import grails.gorm.transactions.Transactional
+import org.apache.http.HttpStatus
+import org.json.JSONObject
+import org.pih.warehouse.component.api.auth.AuthApiWrapper
+import org.pih.warehouse.component.base.ApiSpec
+import spock.lang.Ignore
+import spock.lang.Shared
+
+class AuthApiSpec extends ApiSpec {
+
+    @Shared
+    AuthApiWrapper authApiWrapper
+
+    @Transactional
+    void setupData() {
+        // The base class already creates a test location so nothing to do here.
+    }
+
+    void "login successful"() {
+        given:
+        JSONObject body = buildAuthRequestBody(username, password, location.id)
+
+        expect:
+        authApiWrapper.login(body, HttpStatus.SC_OK)
+    }
+
+    void "login fails with invalid username"() {
+        given:
+        JSONObject body = buildAuthRequestBody("invalid", password, location.id)
+
+        expect:
+        authApiWrapper.login(body, HttpStatus.SC_UNAUTHORIZED)
+    }
+
+    void "login fails with invalid password"() {
+        given:
+        JSONObject body = buildAuthRequestBody(username, "invalid", location.id)
+
+        expect:
+        authApiWrapper.login(body, HttpStatus.SC_UNAUTHORIZED)
+    }
+
+    @Ignore("This actually succeeds, even though the location doesn't exist. Re-enable this test when that's fixed.")
+    void "login fails with invalid location"() {
+        given:
+        JSONObject body = buildAuthRequestBody(username, password, INVALID_ID)
+
+        expect:
+        authApiWrapper.login(body, HttpStatus.SC_UNAUTHORIZED)
+    }
+
+    JSONObject buildAuthRequestBody(String username, String password, String location) {
+        return new JSONObject()
+                .put("username", username)
+                .put("password", password)
+                .put("location", location)
+    }
```

Comment:
This looks like it can be some kind of a generic function that can be used in other tests to convert `Map` to `JsonObject`
some kind of a `JSONObject bodyBuilder(Map data) { ... }` wdyt? :thinking:

Reply:
@drodzewicz JSONObject already has a constructor that takes a map as an argument. For request bodies that are relatively small, this version might be ok, but for bigger I think it might be better to build from a map as you suggested. But I am not sure if this is something that we have to tackle now.

Reply:
A JSONObject is already essentially a map from what I can see so I'm not sure what advantages we'd get working with maps then converting them to JSONObjects.

If we were using request DTOs (data transfer objects) then we could use those directly here to build the request object, but since the controller just parses the request object to JSONObject, this is the best we can do for now